File: Binary Search I/python/Day 4/smallest_char_greater_than_target.py
class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        '''
         Simple binary search but don't return if value found
         a) if value found go to right
         b) at the end s will be reffering the desired char
         c) if s is out of index than return first char
        '''
        s=0
        e=len(letters)-1
        while(s<=e):
            m=s+(e-s)//2
            if(letters[m]>target):
                e=m-1
            else:
                s=m+1
        if(s>=len(letters)):
            return letters[0]
        return letters[s]

File: Binary Search I/cpp/Day 4/SmallestCharGreaterThanTarget.cpp
class Solution {
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
        /**
        Simple binary search but don't return if value found
         a) if value found go to right
         b) at the end s will be reffering the desired char
         c) if s is out of index than return first char
         */
         int s=0;
        int e=letters.size()-1;
        while(s<=e){
            int mid=(s+e)/2;
            if(letters[mid]>target){
                e=mid-1;
            }else{
                s=mid+1;
            }
        }
        if(s>=letters.size()){
            return letters[0];
        }
        return letters[s];
    }
};